[Hazmat Shipping] Handle hazmat category selection

ThomazFB opened the pull request:
Closes #10584

Why
==========
During the Hazmat shipping declaration, it's mandatory to describe in which category the hazardous material falls in. So whenever the Hazmat section is expanded, we should allow the merchant to select the correct category.

This PR introduces the Hazmat category selection UI interaction, but still with no integration with models and network operations. So when we leave the package details view, the selection is lost.

How
==========
Introduces the hazmat category selection flow that's not connected to no models or network operation yet by defining a Button layout containing a flexible area to allow displaying which hazmat category is currently selected, or asking the merchant to select one.

Screen Capture
==========
https://github.com/woocommerce/woocommerce-ios/assets/5920403/a7908a53-ee29-4d76-a1fd-c5d055160f45

How to Test
==========
1. Open the order details of an order eligible for Shipping label creation
2. Start the shipping label creation flow until you reach the Package details section
3. Inside the package details view, make sure the Contains hazardous materials toggle is visible and interactable
4. Make sure the toggle expands and collapses a basic view that will serve as the main control options for the HAZMAT declaration
5. Hit the `Hazardous material category` section, make sure it works as expected for a selection field, and allow selecting or changing the category any time.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Replace TitleAndValueRow with a custom buttom

ThomazFB committed:
Adjust Hazmat selection button alignment

ThomazFB committed:
Finish hazmat category selection button styling

ThomazFB committed:
Define basic HazmatCategory enum in Single Package view

ThomazFB committed:
Define Hazmat Category selection list creation

ThomazFB committed:
Bind hazmat category selection changes to hazmat section

ThomazFB committed:
Create ShippingLabelHazmatCategory Type

ThomazFB committed:
Add all hazmat categories

ThomazFB committed:
Add all hazmat categories localization

ThomazFB committed:
Finish all localization mapping inside Hazmat categories

ThomazFB committed:
Replace old usage of HazmatCategory enum with new one

ThomazFB committed:
Merge branch 'issue/10551-create-hazmat-checkbox' into issue/10584-handle-hazmat-category-selection

ThomazFB committed:
Fix line length issues

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Fix UI incorrect placements after merge

ThomazFB committed:
Remove trailing whitespaces

ThomazFB committed:
Add disclosure indicator to Hazmat category selection

ThomazFB committed:
Fix hound issues

ThomazFB committed:
Remove the none category from the selection list during the Hazmat flow

ThomazFB committed:
Merge branch 'issue/10551-create-hazmat-checkbox' into issue/10584-handle-hazmat-category-selection

ThomazFB committed:
Port hazmat selection changes to SinglePackageHazmatDeclaration

ThomazFB committed:
Remove unnecessary state field from SinglePackage view

ThomazFB committed:
Merge branch 'issue/10551-create-hazmat-checkbox' into issue/10584-handle-hazmat-category-selection

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Multi-package/ShippingLabelHazmatCategory.swift:
```diff
@@ -0,0 +1,161 @@
+    case none
+    case airEligibleEthanol
+    case class1
+    case class3
```
just wondering, there is no class 2? 😆

ThomazFB replied:
Weird, right? 😆 I also had to double check the [easypost API docs](https://www.easypost.com/docs/api) to make sure there's no class 2.
<img width="446" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/5920403/0b290040-44b1-46c3-a191-c2db1c3f09cf">

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Multi-package/ShippingLabelSinglePackageViewModel.swift:
```diff
@@ -124,6 +128,10 @@ final class ShippingLabelSinglePackageViewModel: ObservableObject, Identifiable
 
+    /// The group of selectable Hazardous materials that can be selected
+    ///
+    let selectableHazmatCategories = ShippingLabelHazmatCategory.allCases.filter { $0 != .none }
```
nit: might be worth having a test case to ensure the categories don't include `none`

ThomazFB replied:
Indeed! Added the unit test for this scenario in https://github.com/woocommerce/woocommerce-ios/pull/10585/commits/a3044963904b64b5b4ae0953d76b56e494800852 and also added another one to make sure the ViewModel always start with the `.none` selected in https://github.com/woocommerce/woocommerce-ios/pull/10585/commits/820e5df1084208028abe3234660b7bbf531a52a8, covering the entire category scenario.

jaclync reviewed (approved):
Looks good! 🚢 

Non-blocking: the alignment looks a bit funny

<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/06123e64-8f36-4fb8-bba9-25ce7d1bf765" width="300" />

ThomazFB committed:
Merge branch 'trunk' into issue/10584-handle-hazmat-category-selection

ThomazFB committed:
Fix SelectableItemRow text alignment for large content

ThomazFB committed:
Add unit test for selectable hazmat categories

ThomazFB committed:
Add unit test coverage for the default hazmat category

ThomazFB commented:
@jaclync Thank you for the review! I addressed all your suggestions and the PR is ready for a second round! I just want to check with you about the Selection list text alignment:

I also found odd the alignment like that, and previously verified that just adding the `multilineTextAlignment(.leading)` inside the `SelectableItemRow` fixes the problem. What stopped me was the fear of changing something like this in a component that is used in multiple points of the app, and I don't know if the `multilineTextAlignment` have any caveats I'm not aware about. Do you think it's safe to proceed? I submitted the fix in https://github.com/woocommerce/woocommerce-ios/pull/10585/commits/5c7aea0e892069c4ee55cc6f16dccfda2e1af33d.

jaclync commented:
> I also found odd the alignment like that, and previously verified that just adding the multilineTextAlignment(.leading) inside the SelectableItemRow fixes the problem. What stopped me was the fear of changing something like this in a component that is used in multiple points of the app, and I don't know if the multilineTextAlignment have any caveats I'm not aware about. Do you think it's safe to proceed? I submitted the fix in https://github.com/woocommerce/woocommerce-ios/commit/5c7aea0e892069c4ee55cc6f16dccfda2e1af33d.

Yea, this change affects other screens and it's worth testing these affected views. I tested the SwiftUI views that include `SelectableItemRow` (`ShippingLabelPackageList`, `ShippingLabelServicePackageList`, `ShippingLabelPaymentMethods`) and the multiline text looks good with the change in these views 👍 Feel free to merge it!
